scipy/stats/_hypotests.py

```diff
@@ -663,3 +663,272 @@ def somersd(x, y=None):
+    >>> res.statistic
+    1.894...
+    >>> res.pvalue
+    0.03407...
+
+    Then, the probability of obtaining test results at least as extreme as the
+    data observed above is around 3.4%. Since our p-value is under our
+    significance level defined above, we reject :math:`H_0`. The vaccine has
+    a positive effect and we have :math:`p_1 \leq p_2`; the probability of
+    contracting the virus if we have been vaccinated is lower than if
+    we are not.
+
+    We can compare with Fisher's exact test which produces an exact
+    p-value of 6.4% :
+
+    >>> _, pvalue = stats.fisher_exact([[7, 12], [8, 3]], alternative="less")
+    >>> pvalue
+    0.0640...
+
+    With the same threshold probability of 5%, `fisher_exact` is accepting
+    :math:`H_0` while `barnard_exact` is rejecting it. As stated in [1]_,
+    Barnard's test is uniformly more powerful than Fisher exact test.
+    """
+    if num_it <= 0:
+        raise ValueError(
+            "Number of iterations `num_it` must be strictly positive, "
+            f"found {num_it!r}"
+        )
+
+    c = np.asarray(table, dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0):
+        # If both values in column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return BarnardExactResult(np.nan, 1.0)
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
+
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1)
+
+    p1, p2 = x1 / total_c1, x2 / total_c2
+
+    if pooled:
+        p = (x1 + x2) / (total_c1 + total_c2)
+        var_p1_p2 = p * (1 - p) * (1 / total_c1 + 1 / total_c2)
+    else:
+        var_p1_p2 = p1 * (1 - p1) / total_c1 + p2 * (1 - p2) / total_c2
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide="ignore", invalid="ignore"):
+        TX = np.divide((p1 - p2), np.sqrt(var_p1_p2))
+
+    TX[p1 == p2] = 0  # Removing NaN values
+
+    TX_obs = TX[c[0, 0], c[0, 1]]
+
+    if alternative == "two-sided":
+        idx = np.abs(TX) >= abs(TX_obs)
+    elif alternative == "less":
+        idx = TX <= TX_obs
+    elif alternative == "greater":
+        idx = TX >= TX_obs
+    else:
+        msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            f" found {alternative!r}"
+        )
+        raise ValueError(msg)
+
+    p_value = _binomial_maximisation_of_p_value_with_nuisance_param(
+        total_c1, total_c2, idx, num_it
+    )
+    return BarnardExactResult(TX_obs, p_value)
+
+
+def _binomial_maximisation_of_p_value_with_nuisance_param(
+    total_c1, total_c2, idx, num_it
+):
+    """
+    Maximisation of the pvalue in respect of a nuisance parameter considering
+    a 2x2 sample space.
+    """
```

Comment:
I would add some details either along the code of here about what this is doing exactly in order to follow.

Reply:
Should I really write a long doc as `barnard_exact`, since this is a private function which is only used in barnard and possibly in Boschloo ? I thought that the description was already quite revealing. But if you think it is necessary, I will try to do something